app/src/main/java/com/mapbox/services/android/navigation/testapp/activity/OffboardRouterActivity.kt

```diff
@@ -0,0 +1,257 @@
+            val locationComponent = mapboxMap.locationComponent
+            val options = LocationComponentActivationOptions.builder(this, style).build()
+            locationComponent.activateLocationComponent(options)
+            locationComponent.renderMode = RenderMode.GPS
+            locationComponent.isLocationComponentEnabled = false
+            navigationMapRoute = NavigationMapRoute(navigation, mapView, mapboxMap)
+            Snackbar.make(findViewById(R.id.container), "Tap map to place waypoint", Snackbar.LENGTH_LONG).show()
+            locationEngine = ReplayRouteLocationEngine()
+            newOrigin()
+        }
+    }
+
+    override fun onMapClick(point: LatLng): Boolean {
+        when {
+            destination == null -> {
+                destination = Point.fromLngLat(point.longitude, point.latitude)
+                mapboxMap?.addMarker(MarkerOptions().position(point))
+                calculateRoute()
+            }
+            waypoint == null -> {
+                waypoint = Point.fromLngLat(point.longitude, point.latitude)
+                mapboxMap?.addMarker(MarkerOptions().position(point))
+                calculateRoute()
+            }
+            else -> {
+                Toast.makeText(this, "Only 2 waypoints supported", Toast.LENGTH_LONG).show()
+                mapboxMap?.let { clearMap(it) }
+            }
+        }
+        return false
+    }
+
+    private fun clearMap(map: MapboxMap) {
+        map.clear()
+        route = null
+        destination = null
+        waypoint = null
+        navigationMapRoute.updateRouteVisibilityTo(false)
+        navigationMapRoute.updateRouteArrowVisibilityTo(false)
+    }
+
+    private fun onNewLocationClick() {
+        newOrigin()
+    }
+
+    @SuppressLint("MissingPermission")
+    private fun calculateRoute() {
+        locationEngine.getLastLocation(object : LocationEngineCallback<LocationEngineResult> {
+            override fun onSuccess(result: LocationEngineResult) {
+                findRouteWith(result)
+            }
+
+            override fun onFailure(exception: Exception) {
+                Timber.e(exception)
+            }
+        })
+    }
+
+    private fun findRouteWith(result: LocationEngineResult) {
+        val userLocation = result.lastLocation
+        if (userLocation == null) {
+            Timber.d("calculateRoute: User location is null, therefore, origin can't be set.")
+            return
+        }
+        val origin = Point.fromLngLat(userLocation.longitude, userLocation.latitude)
+        destination?.let { destinationPoint ->
+            if (TurfMeasurement.distance(origin, destinationPoint, TurfConstants.UNIT_METERS) < 50) {
+                return
+            }
+            val waypoints = mutableListOf(waypoint).filterNotNull()
+            val offboardRouter = MapboxOffboardRouter(this, Utils.getMapboxAccessToken(this))
+            directionsSession = MapboxDirectionsSession(
```

Comment:
Can we use one direction session that maintains state instead of creating a new one on each click?

Reply:
Yeah, you are right! I created this example before we remove `requestRoutes()` from `MapboxDirectionsSession#init`... so yeah, absolutely, we need to refactor example a bit)
Very good catch, man!